src/irr/asset/CImageWriterOpenEXR.cpp

```diff
@@ -0,0 +1,159 @@
+namespace IMF = OPENEXR_IMF_NAMESPACE;
+namespace IMATH = IMATH_NAMESPACE;
+
+namespace irr
+{
+	namespace asset
+	{
+		using namespace IMF;
+		using namespace IMATH;
+
+		constexpr uint8_t availableChannels = 4;
+
+		template<typename ilmType>
+		void createAndWriteImage(std::array<ilmType*, availableChannels>& pixelsArrayIlm, const asset::ICPUImage* image, const char* fileName)
+		{
+			auto getIlmType = [&]()
+			{
+				if (image->getCreationParameters().format == EF_R16G16B16A16_SFLOAT)
+					return PixelType::HALF;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_SFLOAT)
+					return PixelType::FLOAT;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_UINT)
+					return PixelType::UINT;
+			};
+
+			Header header(image->getCreationParameters().extent.width, image->getCreationParameters().extent.height);
+			const PixelType pixelType = getIlmType();
+			FrameBuffer frameBuffer;
+
+			const auto& width = image->getCreationParameters().extent.width;
+			const auto& height = image->getCreationParameters().extent.height;
+			const auto pitch = image->getRegions().begin()->bufferRowLength;
+
+			for (uint8_t channel = 0; channel < availableChannels; ++channel)
+				pixelsArrayIlm[channel] = _IRR_NEW_ARRAY(ilmType, width * height);
+
+			for (uint64_t yPos = 0; yPos < height; ++yPos)
+				for (uint64_t xPos = 0; xPos < width; ++xPos)
+				{
+					const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * pitch * availableChannels) + (xPos * availableChannels);
+					const uint64_t ptrStyleIlmShiftToDataChannelPixel = (yPos * width) + xPos;
+
+					for (uint8_t channelIndex = 0; channelIndex < availableChannels; ++channelIndex)
+					{
+						ilmType channelPixel = *(reinterpret_cast<const ilmType*>(image->getBuffer()->getPointer()) + ptrStyleEndShiftToImageDataPixel + channelIndex);
+						*(pixelsArrayIlm[channelIndex] + ptrStyleIlmShiftToDataChannelPixel) = channelPixel;
+					}
+				}
+
+			constexpr std::array<char*, availableChannels> rgbaSignatureAsText = { "R", "G", "B", "A" };
+			for (uint8_t channel = 0; channel < rgbaSignatureAsText.size(); ++channel)
+			{
+				header.channels().insert(rgbaSignatureAsText[channel], Channel(pixelType));
+				frameBuffer.insert
+				(
+					rgbaSignatureAsText[channel],                                                                // name
+					Slice(pixelType,                                                                             // type
+					(char*) pixelsArrayIlm[channel],                                                             // base
+					sizeof(*pixelsArrayIlm[channel]) * 1,                                                        // xStride
+					sizeof(*pixelsArrayIlm[channel]) * image->getCreationParameters().extent.width)              // yStride
+				);
+			}
+
+			OutputFile file(fileName, header);
+			file.setFrameBuffer(frameBuffer);
+			file.writePixels(image->getCreationParameters().extent.height);
```

Comment:
you're only writing 720 pixels of a 1280x720 image here, just saying.

Reply:
It's good, the parameter of this specifies how many scan lines it should write
if I were to put there 1, it would save 1280 x 1 image

because above I describe the end layout, it saves it correctly

Reply:
ah ok. cool